docs/src/integrity/integrity.rst

```diff
@@ -0,0 +1,60 @@
+***********************
+Computational Integrity
+***********************
+
+The integrity module of Opaque ensures that the untrusted job driver hosted on the cloud service schedules tasks in the manner computed by Spark's Catalyst query optimizer. 
+Opaque runs on Spark, which utilizes data partitioning to speed up computation. 
+Specifically, Catalyst will compute a physical query plan for a given dataframe query and delegate Spark workers (run on enclaves) to compute Spark SQL operations on data partitions. 
+Each of these individual units is trusted, but the intermediary steps in which the units communicate is controlled by the job driver, running as untrusted code in the cloud. 
+The integrity module will detect if the job driver has deviated from the query plan computed by Catalyst.
+
+Overview
+--------
+The main idea behind integrity support is to tag each step of computation with a MAC, attached by the enclave worker when it has completed its computation. 
+All MACs received by all previous enclave workers are logged. In the end, these MACs are compared and reconstructed into a graph. 
+This graph is compared to that computed by Catalyst. 
+If the graphs are isomorphic, then no tampering has occurred. 
+Else, the result of the query returned by the cloud is rejected.
+
+Implementation
+--------------
+Two main extensions were made to support integrity - one in enclave code, and one in the Scala client application.
+
+Enclave Code
+^^^^^^^^^^^^
+In the enclave code (C++), modifications were made to the ``FlatbuffersWriters.cpp`` file. 
```

Comment:
Modifications were also made to `FlatbuffersReaders` for the during-execution integrity checks: checking whether all blocks that were outputted from the previous ecall indeed were received by the subsequent ecall.

Reply:
Can you also add a section about the during-execution checks and post-verification checks?

Reply:
Can you elaborate on this? I think I wrote a little about the post verification checks in the part about the "Scala / Job Verification Engine Code" which outlines the reconstruction of the executed and expected. Do you want me to explain more in this section?

Reply:
Yeah I think it'd be good to either have a separate section or to add to the "Overview" section a bit about the during-execution integrity checks and the post-verification checks, i.e. say that as part of integrity we perform checks during execution and post-execution. In particular, maybe we can talk about what each is checking for.

Reply:
Added a little more detail in commit 2db9389